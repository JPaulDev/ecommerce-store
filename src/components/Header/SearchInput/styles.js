import styled from 'styled-components';

export const Form = styled.form`
  justify-self: center;
  position: relative;
  width: 100%;

  &::after {
    border-radius: 9px;
    box-shadow: 0 0 0 3px var(--color-orange-1);
    content: '';
    height: 100%;
    left: 0;
    opacity: ${({ isFocused }) => (isFocused ? '1' : '0')};
    pointer-events: none;
    position: absolute;
    transition: opacity 150ms ease-in-out;
    width: 100%;
  }

  @media screen and (min-width: 545px) {
    max-width: 420px;
  }

  @media screen and (max-width: 929px) {
    grid-column: 1/3;
    grid-row: 2/3;
    margin-top: 15px;
  }
`;

export const Input = styled.input`
  border-radius: 8px;
  color: var(--color-grey-2);
  font-size: 0.94rem;
  font-weight: 500;
  min-height: 44px;
  outline: none;
  padding-left: 10px;
  padding-right: 65px;
  width: 100%;

  &::placeholder {
    font-size: 0.94rem;
  }
`;

export const Button = styled.button`
  align-items: center;
  background-color: var(--color-orange-1);
  border-radius: 8px;
  cursor: pointer;
  display: flex;
  height: 100%;
  justify-content: center;
  padding-left: 3px;
  position: absolute;
  right: 0;
  top: 0;
  width: 55px;
`;
