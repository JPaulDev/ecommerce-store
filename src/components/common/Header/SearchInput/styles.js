import styled, { css } from 'styled-components';

export const Form = styled.form`
  justify-self: center;
  position: relative;
  width: 100%;

  &::after {
    border-radius: var(--border-radius-xl);
    box-shadow: 0 0 0 3px var(--color-orange-1);
    content: '';
    height: 100%;
    left: 0;
    opacity: ${({ isFocused }) => (isFocused ? '1' : '0')};
    pointer-events: none;
    position: absolute;
    transition: opacity 150ms ease-in-out;
    width: 100%;
  }

  @media screen and (min-width: 545px) {
    max-width: 420px;
  }

  @media screen and (max-width: 934px) {
    grid-column: 1 / 3;
    grid-row: 2 / 3;
    margin-top: 15px;
  }
`;

export const Input = css`
  border-radius: var(--border-radius-xl);
  color: var(--color-grey-2);
  font-size: var(--font-size-14);
  font-weight: var(--font-weight-medium);
  min-height: 44px;
  outline: none;
  padding-left: 10px;
  padding-right: 65px;
  width: 100%;

  &::-webkit-search-cancel-button {
    display: none;
  }
`;

export const Button = styled.button`
  align-items: center;
  background-color: var(--color-orange-1);
  border-radius: var(--border-radius-xl);
  cursor: pointer;
  display: flex;
  height: 100%;
  justify-content: center;
  padding-left: 3px;
  position: absolute;
  right: 0;
  top: 0;
  width: 55px;
`;
