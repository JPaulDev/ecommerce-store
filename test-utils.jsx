import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { setupStore } from './src/app/store';
import { UIProvider } from './src/contexts/UIContext';
import theme from './src/styles/theme';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <UIProvider>{children}</UIProvider>
        </Provider>
      </ThemeProvider>
    );
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';
export { renderWithProviders as render };
