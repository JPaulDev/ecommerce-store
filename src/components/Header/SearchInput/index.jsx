import { useState, useRef, useId, useEffect } from 'react';
import { ReactComponent as SearchIcon } from '../../../assets/svg/search.svg';
import * as Styled from './styles';

export default function SearchInput() {
  const [query, setQuery] = useState('');
  const [isFocused, setIsFocused] = useState(false);
  const ref = useRef(null);
  const id = useId();

  const handleChange = (e) => setQuery(e.target.value);

  useEffect(() => {
    const element = ref.current;
    const handleFocus = () => setIsFocused(true);
    const handleBlur = () => setIsFocused(false);

    element.addEventListener('focus', handleFocus);
    element.addEventListener('blur', handleBlur);

    return () => {
      element.removeEventListener('focus', handleFocus);
      element.removeEventListener('blur', handleBlur);
    };
  }, []);

  return (
    <Styled.Form
      role="search"
      action="/search"
      method="get"
      isFocused={isFocused}
    >
      <label htmlFor={id} hidden>
        Search
      </label>
      <Styled.Input
        ref={ref}
        id={id}
        value={query}
        name="searchQuery"
        placeholder="Search PC-Connect.co.uk..."
        type="search"
        autoComplete="off"
        maxLength="50"
        required
        onChange={handleChange}
      />
      <Styled.Button title="Search" type="submit">
        <SearchIcon width={30} height={30} fill="white" />
      </Styled.Button>
    </Styled.Form>
  );
}
