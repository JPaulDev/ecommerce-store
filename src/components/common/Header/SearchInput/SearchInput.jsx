import { useState } from 'react';
import { InputWithLabel } from '../../../ui';
import { Search } from '../../../icons';
import * as Styled from './styles';

export default function SearchInput() {
  const [query, setQuery] = useState('');
  const [isFocused, setIsFocused] = useState(false);

  const handleChangeSearchQuery = (e) => setQuery(e.target.value);

  const handleChangeFocus = () => setIsFocused(!isFocused);

  return (
    <Styled.Form
      role="search"
      action="/search"
      method="get"
      isFocused={isFocused}
    >
      <InputWithLabel
        label="Search:"
        styles={Styled.Input}
        name="searchQuery"
        type="search"
        placeholder="Search PC-Connect.co.uk..."
        maxLength="50"
        required
        aria-required
        onChange={handleChangeSearchQuery}
        onFocus={handleChangeFocus}
        onBlur={handleChangeFocus}
        value={query}
      />
      <Styled.Button title="Search" aria-label="Search" type="submit">
        <Search width={30} height={30} />
      </Styled.Button>
    </Styled.Form>
  );
}
