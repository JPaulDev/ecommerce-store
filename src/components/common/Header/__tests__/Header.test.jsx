import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { render } from '../../../../../test-utils';
import SearchInput from '../index';

function setup(component) {
  return {
    user: userEvent.setup(),
    ...render(component),
  };
}

describe('SearchInput', () => {
  it('should correctly record user input', async () => {
    const { user } = setup(<SearchInput />);
    const input = screen.getByRole('searchbox');

    expect(input).toHaveValue('');

    await user.type(input, 'processors');

    expect(input).toHaveValue('processors');
  });
});
