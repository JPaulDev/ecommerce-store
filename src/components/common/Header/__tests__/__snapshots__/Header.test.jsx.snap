// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders a header with an image, search input and links 1`] = `
.c0 {
  background-color: var(--color-grey-3);
  box-shadow: var(--box-shadow-lg);
  position: relative;
  width: 100%;
  z-index: 100;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: grid;
  grid-template-columns: 1fr 1fr;
  margin: 0 auto;
  max-width: 1280px;
  padding: 20px 15px;
}

.c2 img {
  display: block;
  height: auto;
  max-width: 100%;
}

.c7 {
  -webkit-column-gap: 20px;
  column-gap: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--color-grey-8);
  -webkit-column-gap: 10px;
  column-gap: 10px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: var(--font-size-14);
  position: relative;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color 200ms ease-in-out;
  transition: color 200ms ease-in-out;
}

.c8:hover {
  color: white;
}

.c10 {
  color: white;
  font-size: var(--font-size-16);
  line-height: 1;
  margin-bottom: 5px;
}

.c11 {
  max-width: 115px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c5 {
  border-radius: var(--border-radius-xl);
  color: var(--color-grey-2);
  font-size: var(--font-size-14);
  font-weight: var(--font-weight-medium);
  min-height: 44px;
  outline: none;
  padding-left: 10px;
  padding-right: 65px;
  width: 100%;
}

.c5::-webkit-search-cancel-button {
  display: none;
}

.c3 {
  justify-self: center;
  position: relative;
  width: 100%;
}

.c3::after {
  border-radius: var(--border-radius-xl);
  box-shadow: 0 0 0 3px var(--color-orange-1);
  content: '';
  height: 100%;
  left: 0;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  -webkit-transition: opacity 150ms ease-in-out;
  transition: opacity 150ms ease-in-out;
  width: 100%;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--color-orange-1);
  border-radius: var(--border-radius-xl);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-left: 3px;
  position: absolute;
  right: 0;
  top: 0;
  width: 55px;
}

@media screen and (min-width:590px) {
  .c1 {
    grid-template-columns: 1fr minmax(max-content,1fr);
    width: 97%;
  }
}

@media screen and (min-width:935px) {
  .c1 {
    grid-template-columns: 1fr minmax(270px,1fr) minmax(max-content,1fr);
  }
}

@media screen and (max-width:649px) {
  .c2 {
    width: 155px;
  }
}

@media screen and (min-width:935px) {
  .c7 {
    margin-left: 40px;
  }
}

@media screen and (max-width:589px) {
  .c9 {
    display: none;
  }
}

@media screen and (min-width:545px) {
  .c3 {
    max-width: 420px;
  }
}

@media screen and (max-width:934px) {
  .c3 {
    grid-column: 1 / 3;
    grid-row: 2 / 3;
    margin-top: 15px;
  }
}

<header
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <img
        alt="PC-Connect.co.uk"
        data-nimg="future"
        decoding="async"
        height={24}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={24}
      />
    </div>
    <form
      action="/search"
      className="c3"
      method="get"
      role="search"
    >
      <label
        className="c4"
        htmlFor=":r0:"
      >
        Search:
      </label>
      <input
        aria-required={true}
        autoCapitalize="off"
        autoCorrect="off"
        className="c5"
        id=":r0:"
        maxLength="50"
        name="searchQuery"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search PC-Connect.co.uk..."
        required={true}
        spellCheck="false"
        type="search"
        value=""
      />
      <button
        aria-label="Search"
        className="c6"
        title="Search"
        type="submit"
      >
        <svg
          fill="white"
          height={30}
          viewBox="0 0 21.4 21.4"
          width={30}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.4,15.4A8.5,8.5,0,1,1,17,8.5a8.4,8.4,0,0,1-1.6,4.9L21,19a1.3,1.3,0,0,1,0,2h0a1.3,1.3,0,0,1-2,0l-5.6-5.6ZM8.5,15A6.5,6.5,0,1,0,2,8.5,6.5,6.5,0,0,0,8.5,15Z"
          />
        </svg>
      </button>
    </form>
    <div
      className="c7"
    >
      <button
        aria-labelledby="account-link"
        className="c8"
        title="Account"
      >
        <svg
          fill="var(--color-orange-1)"
          height={30}
          viewBox="0 0 16 15"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="8"
            cy="4"
            r="4"
          />
          <path
            d="M8 9c-6.1 0-8 4-8 4v2h16v-2s-1.9-4-8-4z"
          />
        </svg>
        <div
          className="c9"
          id="account-link"
        >
          <div
            className="c10"
          >
            Account
          </div>
          <div
            className="c11"
          >
            Sign in | Register
          </div>
        </div>
      </button>
      <a
        aria-labelledby="basket-link"
        className="c8"
        href="/basket"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
        title="Basket"
      >
        <svg
          fill="var(--color-orange-1)"
          height={30}
          viewBox="0 0 2048 1792"
          width={34}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1920,768q52.9,0,90.5,37.5T2048,896q0,52.9-37.5,90.5T1920,1024h-15l-115,662q-7.9,46.1-44,76t-82,30H384q-46,0-82-30t-44-76L143,1024H128q-52.9,0-90.5-37.5T0,896q0-52.9,37.5-90.5T128,768ZM485,1568q26-1.9,43.5-22.5c11.7-13.7,16.8-29.2,15.5-46.5l-32-416q-2-25.9-22.5-43.5c-13.7-11.7-29.2-16.8-46.5-15.5s-31.8,8.8-43.5,22.5-16.8,29.2-15.5,46.5l32,416a64.3,64.3,0,0,0,64,59Zm411-64V1088q0-25.9-19-45t-45-19q-25.9,0-45,19t-19,45v416q0,25.9,19,45t45,19q25.9,0,45-19T896,1504Zm384,0V1088q0-25.9-19-45t-45-19q-25.9,0-45,19t-19,45v416q0,25.9,19,45t45,19q25.9,0,45-19T1280,1504Zm352,5,32-416c1.3-17.3-3.8-32.8-15.5-46.5s-26.2-21.2-43.5-22.5-32.8,3.8-46.5,15.5-21.2,26.2-22.5,43.5l-32,416c-1.3,17.3,3.8,32.8,15.5,46.5s26.2,21.2,43.5,22.5h5a64.3,64.3,0,0,0,64-59ZM476,292,383,704H251L352,263q19-88,89-143.5T601,64H768q0-25.9,19-45T832,0h384q25.9,0,45,19t19,45h167q90,0,160,55.5T1696,263l101,441H1665l-93-412q-10.9-44-45.5-72T1447,192H1280q0,26-19,45t-45,19H832q-25.9,0-45-19t-19-45H601q-45,0-79.5,28T476,292Z"
          />
        </svg>
        <div
          className="c9"
          id="basket-link"
        >
          <div
            className="c10"
          >
            Basket
          </div>
          <div>
            Subtotal 
            Â£0.00
          </div>
        </div>
      </a>
    </div>
  </div>
</header>
`;
